File: Modules/VersionManagerModule/Lua/LocalVersionCheckeComp.lua
local LocalVersionCheckeComp = Class.define("LocalVersionCheckeComp",BaseModelComp)


function LocalVersionCheckeComp:Load(MyModel,...)
    self:super(LocalVersionCheckeComp,"Load",MyModel,...);
    self:LoadEnd();
end

 --校验本地环境
function LocalVersionCheckeComp:CheckeLocalVersion(LoadingView,CallBackFunc)
    local localVersionPath = AppConfig.AppAssetBundleAddress .. "/VersionInfo.json"
    local localAssetPath = AppConfig.AppAssetBundleAddress .. "/AssetInfo.json"
    local IsSucc = FilesTools.IsKeepFileOrDirectory(localVersionPath)
    if not IsSucc then --外部资源不存在 进行版本迁移处理
        StartCoroutine(self.AssemblyAsset,self,LoadingView,function()
            local versionstr = FilesTools.ReadFileToStr(localVersionPath)
            local assetinfostr = FilesTools.ReadFileToStr(localAssetPath)
            local LocalVersion = Json.decode(versionstr)
            local LocalAssetInfo = Json.decode(assetinfostr)
            if CallBackFunc then
                CallBackFunc(LocalVersion, LocalAssetInfo)
            end
        end)
    else
        local versionstr = FilesTools.ReadFileToStr(localVersionPath)
        local assetinfostr = FilesTools.ReadFileToStr(localAssetPath)
        local LocalVersion = Json.decode(versionstr)
        local LocalAssetInfo = Json.decode(assetinfostr)
        self:CheckeAppInside(LoadingView,LocalVersion, LocalAssetInfo, CallBackFunc)
    end
end

function LocalVersionCheckeComp:CheckeAppInside(LoadingView,LocalVersion,LocalAssetInfo,CallBackFunc)
    local AppVersion = tonumber(Application.version)
    if LocalVersion.ProVersion < AppVersion  then      --外部资源为上一个版本的资源文件，需要重新覆盖
        print("校验本部资源本版高于外部资源版本 insideVer:"..AppVersion.."  externalVer:"..LocalVersion.ProVersion)
        StartCoroutine(self.AssemblyAsset,self,LoadingView,function()
            local localVersionPath = AppConfig.AppAssetBundleAddress .. "/VersionInfo.json"
            local localAssetPath = AppConfig.AppAssetBundleAddress .. "/AssetInfo.json"
            local versionstr = FilesTools.ReadFileToStr(localVersionPath)
            local assetinfostr = FilesTools.ReadFileToStr(localAssetPath)
            local LocalVersion = Json.decode(versionstr)
            local LocalAssetInfo = Json.decode(assetinfostr)
            if CallBackFunc then
                CallBackFunc(LocalVersion, LocalAssetInfo)
            end
        end)
    else
        if CallBackFunc then
            CallBackFunc(LocalVersion, LocalAssetInfo)
        end
    end
end

function LocalVersionCheckeComp:AssemblyAsset(LoadingView,Func)
    local www =  UnityEngine.Networking.UnityWebRequest.New(AppConfig.GetstreamingAssetsPath .. "/Res.zip")
    local dowle = UnityEngine.Networking.DownloadHandlerBuffer.New()
    www.downloadHandler = dowle
    Yield(www:SendWebRequest())
    if www.error then
        error("读取本地资源文件失败:"..AppConfig.GetstreamingAssetsPath .. "/Res.zip")
    else
        Yield(ZipTools.UnzipFile(dowle.data, AppConfig.AppAssetBundleAddress, AppConfig.ResZipPassword,lgu3d.ZipTools.ZipOpeProgress(function(Describe,Progress)
            LoadingView:UpDateProgress("初次运行解压资源文件"..Describe,Progress)
        end)))  
        if Func then
            Func()
        end
        www:Dispose()
    end
end

--校验本地模块资源信息文件
function LocalVersionCheckeComp:CheckLoadModule(ModuleName,CallBackFunc)
    local localmodulepath = AppConfig.AppAssetBundleAddress .."/"..ModuleName.. "/AssetInfo.json"
    local IsSucc = FilesTools.IsKeepFileOrDirectory(localmodulepath)
    if IsSucc then
        local assetinfostr = FilesTools.ReadFileToStr(localmodulepath)
        local LocalAssetInfo = Json.decode(assetinfostr)
        if CallBackFunc then
            CallBackFunc(nil, LocalAssetInfo)
        end
    else
        local LocalAssetInfo = {AppResInfo={}}
        if CallBackFunc then
            CallBackFunc(nil, LocalAssetInfo)
        end
    end
end

return LocalVersionCheckeComp

File: Modules/VersionManagerModule/Script/LocalVersionCheckeComp.cs
using System;
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using UnityEngine.Networking;

namespace lgu3d
{
    /// <summary>
    /// 版本校验组件
    /// </summary>
    public class LocalVersionCheckeComp : ModelCompBase<VersionManagerModule>
    {

        public override void Load(ModelBase _ModelContorl, params object[] _Agr)
        {
            base.Load(_ModelContorl, _Agr);
            LoadEnd();
        }

        /// <summary>
        /// 校验本地环境
        /// </summary>
        public void CheckeLocalVersion(Action<Dictionary<string, int>, AppModuleAssetInfo> CallBack)
        {
            string localVersionPath = AppConfig.AppAssetBundleAddress + "/VersionInfo.json";
            string localAssetPath = AppConfig.AppAssetBundleAddress + "/AssetInfo.json";
            bool IsSucc = FilesTools.IsKeepFileOrDirectory(localVersionPath);
            if (!IsSucc) //外部资源不存在 进行版本迁移处理
            {
                AssemblyLocalVersion(() =>
                {
                    string versionstr = FilesTools.ReadFileToStr(localVersionPath);
                    string assetinfostr = FilesTools.ReadFileToStr(localAssetPath);
                    Dictionary<string, int> LocalVersion = JsonTools.JsonStrToDictionary<string, int>(versionstr);
                    AppModuleAssetInfo LocalAssetInfo = JsonTools.JsonStrToObject<AppModuleAssetInfo>(assetinfostr);
                    CallBack?.Invoke(LocalVersion, LocalAssetInfo);

                });
            }
            else
            {
                string versionstr = FilesTools.ReadFileToStr(localVersionPath);
                string assetinfostr = FilesTools.ReadFileToStr(localAssetPath);
                Dictionary<string, int> LocalVersion = JsonTools.JsonStrToDictionary<string, int>(versionstr);
                AppModuleAssetInfo LocalAssetInfo = JsonTools.JsonStrToObject<AppModuleAssetInfo>(assetinfostr);
                CheckeAppInside(LocalVersion, LocalAssetInfo, CallBack);
            }

        }

        /// <summary>
        /// 对比App版本
        /// </summary>
        public void CheckeAppInside(Dictionary<string, int> LocalVersion, AppModuleAssetInfo LocalAssetInfo, Action<Dictionary<string, int>, AppModuleAssetInfo> CallBack)
        {
            float AppVersion = float.Parse(Application.version);
            if (LocalVersion["ProVersion"] < AppVersion)        //外部资源为上一个版本的资源文件，需要重新覆盖
            {
                AssemblyLocalVersion(() =>
                {
                    string localVersionPath = AppConfig.AppAssetBundleAddress + "/VersionInfo.json";
                    string localAssetPath = AppConfig.AppAssetBundleAddress + "/AssetInfo.json";
                    string versionstr = FilesTools.ReadFileToStr(localVersionPath);
                    string assetinfostr = FilesTools.ReadFileToStr(localAssetPath);
                    LocalVersion = JsonTools.JsonStrToDictionary<string, int>(versionstr);
                    LocalAssetInfo = JsonTools.JsonStrToObject<AppModuleAssetInfo>(assetinfostr);
                    CallBack?.Invoke(LocalVersion, LocalAssetInfo);
                });
            }
            else
            {
                CallBack?.Invoke(LocalVersion, LocalAssetInfo);
            }
        }

        private void AssemblyLocalVersion(Action CallBack)
        {
            MyModule.InfoOutComp.UpdataView("初次运行解压资源文件", "开始解压资源文件", 0.0f);
            MyModule.StartCoroutine(AssemblyAsset(CallBack));
        }

        IEnumerator AssemblyAsset(Action CallBack)
        {
            UnityWebRequest www = new UnityWebRequest(AppConfig.GetstreamingAssetsPath + "/Res.zip");
            DownloadHandlerBuffer dowle = new DownloadHandlerBuffer();
            www.downloadHandler = dowle;
            yield return www.SendWebRequest();
            if (www.error != null)
                Debug.Log(www.error);
            else
            {
                yield return ZipTools.UnzipFile(dowle.data, AppConfig.AppAssetBundleAddress, AppConfig.ResZipPassword, (string DescribeStr, float Progress) => {
                    MyModule.InfoOutComp.UpdataView("初次运行解压资源文件", DescribeStr, Progress);
                });
                if (CallBack != null)
                {
                    CallBack();
                }
                www.Dispose();
            }
        }


        //校验本地模块资源信息文件
        public void CheckLoadModule(string ModuleName, Action<Dictionary<string, int>, AppModuleAssetInfo> CallBack) {
            string localmodulepath = AppConfig.AppAssetBundleAddress + "/" + ModuleName + "/AssetInfo.json";
            bool IsSucc = FilesTools.IsKeepFileOrDirectory(localmodulepath);
            if (IsSucc)
            {
                string assetinfostr = FilesTools.ReadFileToStr(localmodulepath);
                AppModuleAssetInfo LocalAssetInfo = JsonTools.JsonStrToObject<AppModuleAssetInfo>(assetinfostr);
                CallBack?.Invoke(null, LocalAssetInfo);
            }
            else {
                CallBack?.Invoke(null, new AppModuleAssetInfo());
            }
        }
    }
}
